File: tools/conformance/src/validators/typescript/json/validator.ts
import { parse } from "jsonc-parser";
import mustache from "mustache";

import { validateComments } from "./comments";

import type { ConformanceError } from "../types";

type JsonValue =
  | boolean
  | JsonValue[]
  | null
  | number
  | string
  | { [key: string]: JsonValue };

/**
 * Validates that a generated JSON or JSONC file is a structural superset of
 * its Mustache template and that all template comments are preserved in the
 * instance.
 *
 * **Structural check**: The template is rendered with Mustache then parsed with
 * `jsonc-parser`. Both the rendered template and the instance are compared with
 * `just-diff`. Missing keys (`remove`) and changed values (`replace`) produce
 * errors; extra keys in the instance (`add`) are allowed (superset semantics).
 *
 * **Comment check**: All JSONC comments in the rendered template must appear in
 * the instance in the same relative order. TODO comments match loosely; all
 * others must match exactly.
 */
export function validateJsonConformance(args: {
  data: Record<string, unknown>;
  filename: string;
  instance: string;
  template: string;
}): {
  errors: ConformanceError[];
} {
  const { data, instance, template } = args;

  const renderedTemplate = mustache.render(template, data);

  // Structural validation
  const templateObject = parse(renderedTemplate) as JsonValue;
  const instanceObject = parse(instance) as JsonValue;

  const structuralErrors = validateDepthFirstSearch(
    templateObject,
    instanceObject,
  );

  // Comment validation
  const commentErrors = validateComments({
    instanceText: instance,
    templateText: renderedTemplate,
  });

  return { errors: [...structuralErrors, ...commentErrors] };
}

/**
 * Formats a path array as a readable string (e.g. `["a", "b", 0]` → `"a.b[0]"`).
 */
function formatPath(path: (number | string)[]): string {
  return path.reduce<string>((accumulator, segment) => {
    if (typeof segment === "number")
      return `${accumulator}[${String(segment)}]`;
    return accumulator === "" ? segment : `${accumulator}.${segment}`;
  }, "");
}

function isJsonObject(value: JsonValue): value is Record<string, JsonValue> {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}

/**
 * Recursively compares `template` against `instance`, returning structured
 * `ConformanceError` objects for any missing keys or mismatched values. Extra
 * keys in the instance are silently allowed (superset semantics).
 */
function validateDepthFirstSearch(
  template: JsonValue,
  instance: JsonValue,
  path: (number | string)[] = [],
): ConformanceError[] {
  if (Array.isArray(template) && Array.isArray(instance)) {
    return template.flatMap((item, index) => {
      const itemPath = formatPath([...path, index]);
      return index >= instance.length
        ? [
            {
              errorType: "code" as const,
              fix: `Add the missing array element at index ${String(index)} (path: "${itemPath}") to the instance file.`,
              instancePath: itemPath,
              language: "json" as const,
              message: `Missing required key: "${itemPath}"`,
              templatePath: itemPath,
            },
          ]
        : validateDepthFirstSearch(item, instance[index] ?? null, [
            ...path,
            index,
          ]);
    });
  }

  if (isJsonObject(template) && isJsonObject(instance)) {
    return Object.keys(template).flatMap((key) => {
      const keyPath = formatPath([...path, key]);
      return key in instance
        ? validateDepthFirstSearch(
            template[key] ?? null,
            instance[key] ?? null,
            [...path, key],
          )
        : [
            {
              errorType: "code" as const,
              fix: `Add the missing key "${keyPath}" to the instance file.`,
              instancePath: keyPath,
              language: "json" as const,
              message: `Missing required key: "${keyPath}"`,
              templatePath: keyPath,
            },
          ];
    });
  }

  if (template !== instance) {
    const currentPath = formatPath(path);
    return [
      {
        actual: JSON.stringify(instance),
        errorType: "code" as const,
        expected: JSON.stringify(template),
        fix: `Change the value at "${currentPath}" in the instance file to ${JSON.stringify(template)}.`,
        instancePath: currentPath,
        language: "json" as const,
        message: `Key "${currentPath}": expected ${JSON.stringify(template)}, got ${JSON.stringify(instance)}`,
        templatePath: currentPath,
      },
    ];
  }

  return [];
}

File: tools/conformance/src/modules/validator/python/json/validator.py
"""🔤 JSON/JSONC file conformance validator."""

import json
import re

import chevron

from src.validators.python.json.comments import validate_comments
from src.validators.python.types import ConformanceError

_TOKEN_PATTERN = re.compile(r'("(?:\\.|[^"\\])*")|(//[^\n]*|/\*[\s\S]*?\*/)')


def _strip_comments(text: str) -> str:
    return _TOKEN_PATTERN.sub(lambda m: m.group(1) if m.group(1) else "", text)


def _format_path(path: list) -> str:
    result = ""
    for segment in path:
        if isinstance(segment, int):
            result = f"{result}[{segment}]"
        elif result == "":
            result = segment
        else:
            result = f"{result}.{segment}"
    return result


def _validate_dfs(template, instance, path=None) -> list[ConformanceError]:
    if path is None:
        path = []
    errors = []
    if isinstance(template, list) and isinstance(instance, list):
        for template_item in template:
            current_path = _format_path(path)
            if isinstance(template_item, (bool, int, float, str, type(None))):
                if template_item not in instance:
                    item_path = (
                        f"{current_path}[{json.dumps(template_item)}]"
                        if current_path
                        else f"[{json.dumps(template_item)}]"
                    )
                    errors.append(
                        ConformanceError(
                            error_type="code",
                            language="json",
                            message=f'Missing required value: "{item_path}"',
                            instance_path=item_path,
                            template_path=item_path,
                            fix=f'Add the missing value at "{item_path}" to the instance file.',
                        )
                    )
            else:
                if not instance:
                    errors.append(
                        ConformanceError(
                            error_type="code",
                            language="json",
                            message=f'Missing required value: "{current_path}"',
                            instance_path=current_path,
                            template_path=current_path,
                            fix=f'Add the missing value at "{current_path}" to the instance file.',
                        )
                    )
                else:
                    candidate_errors = []
                    for i, instance_item in enumerate(instance):
                        candidate_errors.append(
                            _validate_dfs(template_item, instance_item, [*path, i])
                        )
                    best = min(candidate_errors, key=len)
                    errors.extend(best)
    elif isinstance(template, dict) and isinstance(instance, dict):
        for key in template:
            key_path = _format_path([*path, key])
            if key in instance:
                errors.extend(_validate_dfs(template[key], instance[key], [*path, key]))
            else:
                errors.append(
                    ConformanceError(
                        error_type="code",
                        language="json",
                        message=f'Missing required key: "{key_path}"',
                        instance_path=key_path,
                        template_path=key_path,
                        fix=f'Add the missing key "{key_path}" to the instance file.',
                    )
                )
    elif template != instance:
        current_path = _format_path(path)
        errors.append(
            ConformanceError(
                error_type="code",
                language="json",
                message=f'Key "{current_path}": expected {json.dumps(template)}, got {json.dumps(instance)}',
                instance_path=current_path,
                template_path=current_path,
                expected=json.dumps(template),
                actual=json.dumps(instance),
                fix=f'Change the value at "{current_path}" in the instance file to {json.dumps(template)}.',
            )
        )
    return errors


def validate_json_conformance(*, data: dict, filename: str, instance: str, template: str) -> dict:
    rendered = chevron.render(template, data)
    template_obj = json.loads(_strip_comments(rendered))
    instance_obj = json.loads(_strip_comments(instance))
    structural_errors = _validate_dfs(template_obj, instance_obj)
    comment_errors = validate_comments(template_text=rendered, instance_text=instance)
    return {"errors": [*structural_errors, *comment_errors]}
